File: models/daydata.go
package models

import (
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// DayData contains the basic data for a given day
type DayData struct {
	gorm.Model

	Date datatypes.Date `gorm:"column:Date"`

	CashDeposit float64 `gorm:"column:CashDeposit"`
	DebitCard   float64 `gorm:"column:DebitCard"`
	MasterCard  float64 `gorm:"column:MasterCard"`
	Visa        float64 `gorm:"column:Visa"`
	Amex        float64 `gorm:"column:Amex"`

	CreditSales    float64 `gorm:"column:CreditSales"`
	GiftCardRedeem float64 `gorm:"column:GiftCardRedeem"`
	SubwayCaters   float64 `gorm:"column:SubwayCaters"`
	PayPal         float64 `gorm:"column:PayPal"`
	SkipTheDishes  float64 `gorm:"column:SkipTheDishes"`
	DoorDash       float64 `gorm:"column:DoorDash"`
	PettyCash      float64 `gorm:"column:PettyCash"`

	Tips                 float64 `gorm:"column:Tips"`
	HST                  float64 `gorm:"column:HST"`
	BottleDeposit        float64 `gorm:"column:BottleDeposit"`
	NetSales             float64 `gorm:"column:NetSales"`
	CreditSalesRedeemed  float64 `gorm:"column:CreditSalesRedeemed"`
	CreditSalesRedeemed2 float64 `gorm:"column:CreditSalesRedeemed2"`
	CreditFood           float64 `gorm:"column:CreditFood"`
	GiftCardSold         float64 `gorm:"column:GiftCardSold"`

	USFunds float64 `gorm:"column:USFunds"`

	//
	//	Control Sheet Information
	//

	HoursWorked    float64 `gorm:"column:HoursWorked"`
	Productivity   float64 `gorm:"column:Productivity"`
	Factor         float64 `gorm:"column:Factor"`
	AdjustedSales  float64 `gorm:"column:AdjustedSales"`
	CustomerCount  int     `gorm:"column:CustomerCount"`
	BreadCredits   float64 `gorm:"column:BreadCredits"`
	BreadOverShort float64 `gorm:"column:BreadOverShort"`

	// holds average of previous 4 weeks
	WeeklyAverage float64 `gorm:"column:WeeklyAverage"`

	Comment string `gorm:"column:Comment"` // comment for the given day
}

// TODO:
//	This structure can be removed along with any code relant on it
//	once the need to convert a django version of the db has passed
//
// Comments contains day id and the comment linked to it
type Comments struct {
	gorm.Model

	LinkedID int    `gorm:"column:LinkedID"`
	Comment  string `gorm:"column:Comment"`
}

// TagInfo simply provides a name to the given tag
type TagList struct {
	gorm.Model

	Tag string `gorm:"column:Tag"`
}

// TagData keeps track of which days have what tags
type TagData struct {
	gorm.Model

	TagID uint `gorm:"column:TagID"` // ID of the TagInfo structure for this entry
	DayID uint `gorm:"column:DayID"`
}

File: backend-rust/src/models/day_data.rs
#![allow(non_snake_case)]
use chrono::NaiveDate;
use diesel::prelude::*;
use diesel::result::Error;
use serde::{Deserialize, Serialize};

#[derive(Queryable, Selectable, AsChangeset)]
#[diesel(table_name=crate::schema::day_data)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
#[derive(Serialize, Deserialize, Clone)]
pub struct DayData {
    pub id: i32,

    // Debit Side
    pub DayDate: NaiveDate,
    pub CashDeposit: i32,
    pub DebitCard: i32,
    pub MasterCard: i32,
    pub Visa: i32,
    pub Amex: i32,
    pub CreditSales: i32,
    pub GiftCardRedeem: i32,
    pub SubwayCaters: i32,
    pub PayPal: i32,
    pub SkipTheDishes: i32,
    pub DoorDash: i32,
    pub UberEats: i32,
    pub PettyCash: i32,

    // Credit Side
    pub Tips: i32,
    pub Hst: i32,
    pub BottleDeposit: i32,
    pub NetSales: i32,
    pub CreditSalesRedeemed: i32,
    pub CreditFood: i32,
    pub GiftCardSold: i32,
    pub USFunds: i32,
    pub CommentData: Option<String>,

    // Control Sheet
    pub HoursWorked: i32,
    pub Productivity: i32,
    pub Factor: i32,
    pub AdjustedSales: i32,
    pub CustomerCount: i32,
    pub BreadCredits: i32,
    pub BreadOverShort: i32,
}

///
/// Have a 2nd copy identical to DayData except for missing id, to allow for
/// autoincrememintg id.  any other fields that don't need to get inserted
/// can also be removed from this struct
///
#[derive(Insertable)]
#[diesel(table_name=crate::schema::day_data)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct DayDataInsert {
    pub DayDate: NaiveDate,
    pub CashDeposit: i32,
    pub DebitCard: i32,
    pub MasterCard: i32,
    pub Visa: i32,
    pub Amex: i32,
    pub CreditSales: i32,
    pub GiftCardRedeem: i32,
    pub SubwayCaters: i32,
    pub PayPal: i32,
    pub SkipTheDishes: i32,
    pub DoorDash: i32,
    pub UberEats: i32,
    pub PettyCash: i32,
    pub Tips: i32,
    pub Hst: i32,
    pub BottleDeposit: i32,
    pub NetSales: i32,
    pub CreditSalesRedeemed: i32,
    pub CreditFood: i32,
    pub GiftCardSold: i32,
    pub USFunds: i32,
    pub CommentData: Option<String>,
    pub HoursWorked: i32,
    pub Productivity: i32,
    pub Factor: i32,
    pub AdjustedSales: i32,
    pub CustomerCount: i32,
    pub BreadCredits: i32,
    pub BreadOverShort: i32,
}

impl DayData {
    pub fn new(date: chrono::NaiveDate) -> Self {
        Self {
            id: -1,
            DayDate: date,
            CashDeposit: 0,
            DebitCard: 0,
            MasterCard: 0,
            Visa: 0,
            Amex: 0,
            CreditSales: 0,
            GiftCardRedeem: 0,
            SubwayCaters: 0,
            PayPal: 0,
            SkipTheDishes: 0,
            DoorDash: 0,
            UberEats: 0,
            PettyCash: 0,
            Tips: 0,
            Hst: 0,
            BottleDeposit: 0,
            NetSales: 0,
            CreditSalesRedeemed: 0,
            CreditFood: 0,
            GiftCardSold: 0,
            USFunds: 0,
            CommentData: None,
            HoursWorked: 0,
            Productivity: 0,
            Factor: 0,
            AdjustedSales: 0,
            CustomerCount: 0,
            BreadCredits: 0,
            BreadOverShort: 0,
        }
    }

    // copy controlsheet data from one object to this one
    // used before updating after a daily sheet import
    pub fn copy_control(&mut self, data: &DayData) {
        self.HoursWorked = data.HoursWorked;
        self.Productivity = data.Productivity;
        self.Factor = data.Factor;
        self.AdjustedSales = data.AdjustedSales;
        self.CustomerCount = data.CustomerCount;
        self.BreadCredits = data.BreadCredits;
        self.BreadOverShort = data.BreadOverShort;
    }

    pub fn insert_or_update(&self, conn: &mut SqliteConnection) -> Result<(), Error> {
        // if data.id == -1, we insert, otherwise we update
        if self.id == -1 {
            // insert
            let data_insert = DayDataInsert::from(self);

            diesel::insert_into(crate::schema::day_data::table)
                .values(&data_insert)
                .execute(conn)?;
        } else {
            // update
            use crate::schema::day_data::dsl::*;

            diesel::update(crate::schema::day_data::table)
                .filter(id.eq(self.id)) // make sure to only update the record we are
                .set(self)
                .execute(conn)?;
        }
        Ok(())
    }
}

impl DayDataInsert {
    pub fn new(date: chrono::NaiveDate) -> Self {
        Self {
            DayDate: date,
            CashDeposit: 0,
            DebitCard: 0,
            MasterCard: 0,
            Visa: 0,
            Amex: 0,
            CreditSales: 0,
            GiftCardRedeem: 0,
            SubwayCaters: 0,
            PayPal: 0,
            SkipTheDishes: 0,
            DoorDash: 0,
            UberEats: 0,
            PettyCash: 0,
            Tips: 0,
            Hst: 0,
            BottleDeposit: 0,
            NetSales: 0,
            CreditSalesRedeemed: 0,
            CreditFood: 0,
            GiftCardSold: 0,
            USFunds: 0,
            CommentData: None,
            HoursWorked: 0,
            Productivity: 0,
            Factor: 0,
            AdjustedSales: 0,
            CustomerCount: 0,
            BreadCredits: 0,
            BreadOverShort: 0,
        }
    }

    // create an insertable entry from the main entry
    pub fn from(data: &DayData) -> Self {
        Self {
            DayDate: data.DayDate,
            CashDeposit: data.CashDeposit,
            DebitCard: data.DebitCard,
            MasterCard: data.MasterCard,
            Visa: data.Visa,
            Amex: data.Amex,
            CreditSales: data.CreditSales,
            GiftCardRedeem: data.GiftCardRedeem,
            SubwayCaters: data.SubwayCaters,
            PayPal: data.PayPal,
            SkipTheDishes: data.SkipTheDishes,
            DoorDash: data.DoorDash,
            UberEats: data.UberEats,
            PettyCash: data.PettyCash,
            Tips: data.Tips,
            Hst: data.Hst,
            BottleDeposit: data.BottleDeposit,
            NetSales: data.NetSales,
            CreditSalesRedeemed: data.CreditSalesRedeemed,
            CreditFood: data.CreditFood,
            GiftCardSold: data.GiftCardSold,
            USFunds: data.USFunds,
            CommentData: data.CommentData.clone(),
            HoursWorked: data.HoursWorked,
            Productivity: data.Productivity,
            Factor: data.Factor,
            AdjustedSales: data.AdjustedSales,
            CustomerCount: data.CustomerCount,
            BreadCredits: data.BreadCredits,
            BreadOverShort: data.BreadOverShort,
        }
    }
}
